src/Rings/mpoly-graded.jl

```diff
@@ -2012,38 +2012,33 @@ function _hilbert_numerator_bayer_stillman(
   #
   # Algorithm 2.6, page 35
   ###########################################################################
-  ret = _hilbert_numerator_trivial_cases(a, weight_matrix, S)
+  ret = _hilbert_numerator_trivial_cases(a, weight_matrix, S, t, oneS)
   ret !== nothing && return ret
 
-  t = gens(S)
-
   # make sure we have lexicographically ordered monomials
-  sort!(a)
+  sort!(a, alg=QuickSort)
 
   # initialize the result 
-  h = one(S) - prod(t[i]^sum([a[1][j]*weight_matrix[i, j] for j in 1:length(a[1])]) for i in 1:length(t))
+  h = oneS - _expvec_to_poly(S, t, weight_matrix, a[1])
+  linear_mons = Vector{Int}()
   for i in 2:length(a)
     J = _divide_by(a[1:i-1], a[i])
-    J1 = Vector{Vector{Int}}()
-    linear_mons = Vector{Int}()
-    for m in J
-      if sum(m) > 1
-        push!(J1, m)
-      else
-        for k in 1:length(m)
-          if m[k] > 0 
-            push!(linear_mons, k)
-            break
-          end
-        end
+    empty!(linear_mons)
```

Comment:
This is also to make sure that we do not need another new allocation, right?

Reply:
exactly